akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -2147,12 +2147,21 @@ private[akka] object TakeWithin {
       override def onPush(): Unit = push(out, grab(in))
       override def onUpstreamFailure(ex: Throwable): Unit = onFailure(ex)
       override def onPull(): Unit = pull(in)
-      def onFailure(ex: Throwable): Unit =
-        if ((maximumRetries < 0 || attempt < maximumRetries) && pf.isDefinedAt(ex)) {
-          switchTo(pf(ex))
-          attempt += 1
+      def onFailure(ex: Throwable): Unit = {
+        import Collect.NotApplied
+        if (maximumRetries < 0 || attempt < maximumRetries) {
+          pf.applyOrElse(ex, NotApplied) match {
+            case NotApplied                                                                => failStage(ex)
+            case source: Source[T @unchecked, _] if source eq Source.empty                 => completeStage()
+            case source: javadsl.Source[T @unchecked, _] if source.asScala eq Source.empty => completeStage()
```

Comment:
In Flow#internalConcat we only look if it is the same instance as Scala DSL empty, are you sure a special case for Java DSL here, and if so, do we need a special case for Java empty there as well?

Reply:
when delegate to scaladsl, `asScala` is not called, so it's needed here.

Reply:
But not for the internalConcat case?

Reply:
> But not for the internalConcat case?

Have not check that part yet...

Reply:
Thanks for looking into that, If it is needed there as well it might be worth extracting into some shared internal logic and use across optimization sites.